src/Data/String/CodePoints.purs

```diff
@@ -0,0 +1,273 @@
+-- | These functions allow PureScript strings to be treated as if they were
+-- | sequences of Unicode code points instead of their true underlying
+-- | implementation (sequences of UTF-16 code units). For nearly all uses of
+-- | strings, these functions should be preferred over the ones in Data.String.
+module Data.String.CodePoints
+  ( module StringReExports
+  , CodePoint()
+  , codePointAt
+  , codePointFromInt
+  , codePointToInt
+  , count
+  , drop
+  , dropWhile
+  , fromCodePointArray
+  , indexOf
+  , indexOf'
+  , lastIndexOf
+  , lastIndexOf'
+  , length
+  , singleton
+  , splitAt
+  , take
+  , takeWhile
+  , toCodePointArray
+  , uncons
+  ) where
+
+import Prelude
+
+import Data.Array as Array
+import Data.Char as Char
+import Data.Maybe (Maybe(Just, Nothing))
+import Data.String as String
+import Data.String.Unsafe as Unsafe
+-- WARN: In order for this module to be a drop-in replacement for Data.String,
+-- this list must be updated to re-export any exports added to Data.String.
+import Data.String (Pattern(..), Replacement(..), charAt, charCodeAt, contains, fromCharArray, joinWith, localeCompare, null, replace, replaceAll, split, stripPrefix, stripSuffix, toChar, toCharArray, toLower, toUpper, trim) as StringReExports
+import Data.Tuple (Tuple(Tuple))
+import Data.Unfoldable (unfoldr)
+
+
+-- | CodePoint is an Int bounded between 0 and 0x10FFFF, corresponding to
+-- | Unicode code points.
+newtype CodePoint = CodePoint Int
+
+derive instance eqCodePoint :: Eq CodePoint
+derive instance ordCodePoint :: Ord CodePoint
+
+-- I would prefer that this smart constructor not need to exist and instead
+-- CodePoint just implements Enum, but the Enum module already depends on this
+-- one. To avoid the circular dependency, we just expose these two functions.
+codePointFromInt :: Int -> Maybe CodePoint
+codePointFromInt n | 0 <= n && n <= 0x10FFFF = Just (CodePoint n)
+codePointFromInt n = Nothing
+
+codePointToInt :: CodePoint -> Int
+codePointToInt (CodePoint n) = n
+
+unsurrogate :: Int -> Int -> CodePoint
```

Comment:
Every use of this function has an `isLead cu0 && isTrail cu1` check beforehand; do you think it might make more sense to include this check inside the `unsurrogate` function and have it return a `Maybe CodePoint`, returning `Nothing` if it is given arguments which don't form a surrogate pair?